File: academic-system-java/src/main/java/org/example/academic/system/view/ConsoleMenu.java
package org.example.academic.system.view;

import org.example.academic.system.controller.AcademicSystemController;
import org.example.academic.system.controller.AuthenticationController;
import org.example.academic.system.exception.AcademicSystemException;
import org.example.academic.system.exception.AuthenticationException;
import org.example.academic.system.exception.AuthorizationException;
import org.example.academic.system.exception.KeyboardInputException;
import org.example.academic.system.model.AcademicClass;
import org.example.academic.system.model.Assessment;
import org.example.academic.system.model.PersistenceType;
import org.example.academic.system.model.Role;
import org.example.academic.system.model.User;

import java.util.List;
import java.util.Scanner;

public class ConsoleMenu {
    private final AcademicSystemController controller;
    private final AuthenticationController authenticationController;
    private final Scanner scanner;

    public ConsoleMenu(AcademicSystemController controller, AuthenticationController authenticationController) {
        this.controller = controller;
        this.authenticationController = authenticationController;
        this.scanner = new Scanner(System.in);
    }

    public void start() {
        boolean running = true;

        while (running) {
            User user = login();

            if (user.getRole() == Role.ADMIN) {
                running = showAdminMenu();
            } else if (user.getRole() == Role.PROFESSOR) {
                running = showProfessorMenu();
            }
        }
    }

    private User login() {
        while (true) {
            System.out.println();
            System.out.println("===== Sistema Academico =====");
            System.out.println();
            String username = readText("Usuario: ");
            String password = readText("Senha: ");

            try {
                User user = authenticationController.login(username, password);
                System.out.println("Login realizado com sucesso.");
                System.out.println("Bem-vindo, " + user.getUsername() + "!");
                return user;
            } catch (AuthenticationException exception) {
                System.out.println("Usuario ou senha invalidos.");
            }
        }
    }

    private boolean showAdminMenu() {
        int option;

        do {
            printAdminMenu();
            try {
                option = readInteger("Escolha uma opcao: ");

                switch (option) {
                    case 1 -> execute(this::registerClass);
                    case 2 -> execute(this::registerAssessment);
                    case 3 -> execute(this::listClasses);
                    case 4 -> execute(this::printClassAssessmentSummaryReport);
                    case 5 -> execute(this::printAssessmentWeightReport);
                    case 6 -> execute(this::configurePersistenceType);
                    case 7 -> execute(this::saveAcademicData);
                    case 8 -> execute(this::printPersistenceConfigurationReport);
                    case 9 -> {
                        authenticationController.logout();
                        System.out.println("Logout realizado com sucesso.");
                        return true;
                    }
                    case 0 -> {
                        System.out.println("Programa encerrado.");
                        return false;
                    }
                    default -> throw new KeyboardInputException("Opcao invalida.");
                }
            } catch (KeyboardInputException exception) {
                System.out.println("Erro de entrada: " + exception.getMessage());
            }
        } while (true);
    }

    private boolean showProfessorMenu() {
        int option;

        do {
            printProfessorMenu();
            try {
                option = readInteger("Escolha uma opcao: ");

                switch (option) {
                    case 1 -> execute(this::registerAssessment);
                    case 2 -> execute(this::listClasses);
                    case 3 -> execute(this::printClassAssessmentSummaryReport);
                    case 4 -> execute(this::printAssessmentWeightReport);
                    case 5 -> {
                        authenticationController.logout();
                        System.out.println("Logout realizado com sucesso.");
                        return true;
                    }
                    case 0 -> {
                        System.out.println("Programa encerrado.");
                        return false;
                    }
                    default -> throw new KeyboardInputException("Opcao invalida.");
                }
            } catch (KeyboardInputException exception) {
                System.out.println("Erro de entrada: " + exception.getMessage());
            }
        } while (true);
    }

    private void printAdminMenu() {
        System.out.println();
        System.out.println("===== Sistema Academico - ADMIN =====");
        System.out.println();
        System.out.println("1 - Cadastrar turma");
        System.out.println("2 - Cadastrar avaliacao");
        System.out.println("3 - Listar turmas");
        System.out.println("4 - Gerar relatorio de avaliacoes por turma");
        System.out.println("5 - Gerar relatorio de peso das avaliacoes");
        System.out.println("6 - Configurar tipo de persistencia");
        System.out.println("7 - Salvar dados academicos");
        System.out.println("8 - Gerar relatorio de configuracao de persistencia");
        System.out.println("9 - Logout");
        System.out.println("0 - Sair");
    }

    private void printProfessorMenu() {
        System.out.println();
        System.out.println("===== Sistema Academico - PROFESSOR =====");
        System.out.println();
        System.out.println("1 - Cadastrar avaliacao");
        System.out.println("2 - Listar turmas");
        System.out.println("3 - Gerar relatorio de avaliacoes por turma");
        System.out.println("4 - Gerar relatorio de peso das avaliacoes");
        System.out.println("5 - Logout");
        System.out.println("0 - Sair");
    }

    private void execute(MenuAction action) {
        try {
            action.run();
        } catch (AcademicSystemException exception) {
            System.out.println("Erro: " + exception.getMessage());
        } catch (AuthorizationException exception) {
            System.out.println("Acesso negado.");
        } catch (KeyboardInputException exception) {
            System.out.println("Erro de entrada: " + exception.getMessage());
        }
    }

    private void registerClass() {
        String code = readText("Codigo da turma: ");
        String title = readText("Titulo da turma: ");

        controller.registerClass(code, title);
        System.out.println("Turma cadastrada com sucesso.");
    }

    private void registerAssessment() {
        String classCode = readText("Codigo da turma: ");
        String assessmentType = readAssessmentType();
        double value = readDouble("Valor da avaliacao: ");
        double weight = readDouble("Peso da avaliacao: ");

        controller.registerAssessment(classCode, assessmentType, value, weight);
        System.out.println("Avaliacao cadastrada com sucesso.");
    }

    private void listClasses() {
        List<AcademicClass> classes = controller.listClasses();

        if (classes.isEmpty()) {
            System.out.println("Nenhuma turma cadastrada.");
            return;
        }

        for (AcademicClass academicClass : classes) {
            System.out.println(academicClass);

            if (academicClass.getAssessments().isEmpty()) {
                System.out.println("  Nenhuma avaliacao cadastrada.");
                continue;
            }

            for (Assessment assessment : academicClass.getAssessments()) {
                System.out.println("  - Tipo: " + formatAssessmentType(assessment)
                        + " | Valor: " + assessment.getValue()
                        + " | Peso: " + assessment.getWeight());
            }
        }
    }

    private String formatAssessmentType(Assessment assessment) {
        return switch (assessment.getType()) {
            case "Exam" -> "Exame";
            case "Practical Assignment" -> "Atividade pratica";
            case "Seminar" -> "Seminario";
            case "Assignment" -> "Trabalho";
            default -> assessment.getType();
        };
    }

    private void printClassAssessmentSummaryReport() {
        System.out.println(controller.generateClassAssessmentSummaryReport());
    }

    private void printAssessmentWeightReport() {
        System.out.println(controller.generateAssessmentWeightReport());
    }

    private void saveAcademicData() {
        controller.saveAcademicData();
        System.out.println("Dados academicos salvos com sucesso.");
    }

    private void printPersistenceConfigurationReport() {
        System.out.println(controller.generatePersistenceConfigurationReport());
    }

    private void configurePersistenceType() {
        System.out.println("Tipo de persistencia:");
        System.out.println("1 - TXT");
        System.out.println("2 - JSON");
        System.out.println("3 - XML");

        while (true) {
            try {
                int option = readInteger("Escolha o tipo de persistencia: ");

                switch (option) {
                    case 1 -> {
                        controller.configurePersistence(PersistenceType.TXT);
                        System.out.println("Persistencia configurada como TXT.");
                        System.out.println("Turmas carregadas: " + controller.listClasses().size());
                        return;
                    }
                    case 2 -> {
                        controller.configurePersistence(PersistenceType.JSON);
                        System.out.println("Persistencia configurada como JSON.");
                        System.out.println("Turmas carregadas: " + controller.listClasses().size());
                        return;
                    }
                    case 3 -> {
                        controller.configurePersistence(PersistenceType.XML);
                        System.out.println("Persistencia configurada como XML.");
                        System.out.println("Turmas carregadas: " + controller.listClasses().size());
                        return;
                    }
                    default -> throw new KeyboardInputException("Tipo de persistencia invalido.");
                }
            } catch (KeyboardInputException exception) {
                System.out.println("Erro de entrada: " + exception.getMessage());
            }
        }
    }

    private String readText(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine();
    }

    private String readAssessmentType() {
        System.out.println("Tipo da avaliacao:");
        System.out.println("1 - Exame");
        System.out.println("2 - Atividade pratica");
        System.out.println("3 - Seminario");
        System.out.println("4 - Trabalho");

        while (true) {
            try {
                int option = readInteger("Escolha o tipo da avaliacao: ");

                switch (option) {
                    case 1 -> {
                        return "exam";
                    }
                    case 2 -> {
                        return "practical";
                    }
                    case 3 -> {
                        return "seminar";
                    }
                    case 4 -> {
                        return "assignment";
                    }
                    default -> throw new KeyboardInputException("Tipo de avaliacao invalido.");
                }
            } catch (KeyboardInputException exception) {
                System.out.println("Erro de entrada: " + exception.getMessage());
            }
        }
    }

    private int readInteger(String prompt) {
        try {
            System.out.print(prompt);
            return Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException exception) {
            throw new KeyboardInputException("Numero inteiro invalido.");
        }
    }

    private double readDouble(String prompt) {
        try {
            System.out.print(prompt);
            String value = scanner.nextLine().trim().replace(',', '.');
            return Double.parseDouble(value);
        } catch (NumberFormatException exception) {
            throw new KeyboardInputException("Numero decimal invalido.");
        }
    }

    private interface MenuAction {
        void run();
    }
}

File: academic-system-python/academic_system/view/console_menu.py
from ..controller.academic_system_controller import AcademicSystemController
from ..controller.authentication_controller import AuthenticationController
from ..exception.academic_system_exception import AcademicSystemException
from ..exception.authentication_exception import AuthenticationException
from ..exception.authorization_exception import AuthorizationException
from ..exception.invalid_menu_option_exception import InvalidMenuOptionException
from ..exception.invalid_numeric_input_exception import InvalidNumericInputException
from ..model.persistence_type import PersistenceType


class ConsoleMenu:
    def __init__(
        self,
        academic_controller: AcademicSystemController,
        authentication_controller: AuthenticationController,
    ):
        self._academic_controller = academic_controller
        self._authentication_controller = authentication_controller

    def start(self) -> None:
        while True:
            user = self._login()
            if user.role.name == "ADMIN":
                should_exit = self._show_admin_menu()
            else:
                should_exit = self._show_professor_menu()
            if should_exit:
                break

    def _login(self):
        while True:
            print("\n===== Login =====")
            username = input("Username: ").strip()
            password = input("Password: ").strip()
            try:
                return self._authentication_controller.login(username, password)
            except AuthenticationException as e:
                print(f"Erro de autenticacao: {e}")

    def _show_admin_menu(self) -> bool:
        while True:
            print("\n===== Menu ADMIN =====")
            print("1. Cadastrar turma")
            print("2. Cadastrar avaliacao")
            print("3. Listar turmas")
            print("4. Relatorio de avaliacoes por turma")
            print("5. Relatorio de peso das avaliacoes")
            print("6. Configurar tipo de persistencia")
            print("7. Salvar dados academicos")
            print("8. Relatorio de configuracao de persistencia")
            print("9. Logout")
            print("0. Sair")
            try:
                option = self._read_int("Opcao: ")
                if option == 1:
                    self._register_class()
                elif option == 2:
                    self._register_assessment()
                elif option == 3:
                    self._list_classes()
                elif option == 4:
                    print(self._academic_controller.generate_class_assessment_summary_report())
                elif option == 5:
                    print(self._academic_controller.generate_assessment_weight_report())
                elif option == 6:
                    self._configure_persistence()
                elif option == 7:
                    self._academic_controller.save_academic_data()
                    print("Dados salvos com sucesso.")
                elif option == 8:
                    print(self._academic_controller.generate_persistence_configuration_report())
                elif option == 9:
                    self._authentication_controller.logout()
                    print("Logout realizado com sucesso.")
                    return False
                elif option == 0:
                    return True
                else:
                    raise InvalidMenuOptionException("Opcao invalida.")
            except (AcademicSystemException, AuthorizationException) as e:
                print(f"Erro: {e}")
            except InvalidMenuOptionException as e:
                print(f"Opcao invalida: {e}")
            except InvalidNumericInputException as e:
                print(f"Entrada invalida: {e}")

    def _show_professor_menu(self) -> bool:
        while True:
            print("\n===== Menu PROFESSOR =====")
            print("1. Cadastrar avaliacao")
            print("2. Listar turmas")
            print("3. Relatorio de avaliacoes por turma")
            print("4. Relatorio de peso das avaliacoes")
            print("5. Logout")
            print("0. Sair")
            try:
                option = self._read_int("Opcao: ")
                if option == 1:
                    self._register_assessment()
                elif option == 2:
                    self._list_classes()
                elif option == 3:
                    print(self._academic_controller.generate_class_assessment_summary_report())
                elif option == 4:
                    print(self._academic_controller.generate_assessment_weight_report())
                elif option == 5:
                    self._authentication_controller.logout()
                    print("Logout realizado com sucesso.")
                    return False
                elif option == 0:
                    return True
                else:
                    raise InvalidMenuOptionException("Opcao invalida.")
            except (AcademicSystemException, AuthorizationException) as e:
                print(f"Erro: {e}")
            except InvalidMenuOptionException as e:
                print(f"Opcao invalida: {e}")
            except InvalidNumericInputException as e:
                print(f"Entrada invalida: {e}")

    def _register_class(self) -> None:
        print("\n--- Cadastrar Turma ---")
        code = input("Codigo: ").strip()
        title = input("Titulo: ").strip()
        self._academic_controller.register_class(code, title)
        print("Turma cadastrada com sucesso.")

    def _register_assessment(self) -> None:
        print("\n--- Cadastrar Avaliacao ---")
        class_code = input("Codigo da turma: ").strip()
        print("Tipos: exam, practical assignment, seminar, assignment")
        assessment_type = input("Tipo: ").strip()
        value = self._read_float("Valor: ")
        weight = self._read_float("Peso: ")
        self._academic_controller.register_assessment(class_code, assessment_type, value, weight)
        print("Avaliacao cadastrada com sucesso.")

    def _list_classes(self) -> None:
        classes = self._academic_controller.list_classes()
        if not classes:
            print("Nenhuma turma cadastrada.")
            return
        for c in classes:
            print(f"  {c.code} - {c.title} ({len(c.assessments)} avaliacao(oes))")

    def _configure_persistence(self) -> None:
        print("\n--- Configurar Persistencia ---")
        print("1. TXT  2. JSON  3. XML")
        option = self._read_int("Opcao: ")
        mapping = {1: PersistenceType.TXT, 2: PersistenceType.JSON, 3: PersistenceType.XML}
        if option not in mapping:
            raise InvalidMenuOptionException("Opcao invalida.")
        self._academic_controller.configure_persistence(mapping[option])
        print("Persistencia configurada com sucesso.")

    def _read_int(self, prompt: str) -> int:
        try:
            return int(input(prompt).strip())
        except ValueError:
            raise InvalidNumericInputException("Digite um numero inteiro.")

    def _read_float(self, prompt: str) -> float:
        try:
            return float(input(prompt).strip().replace(",", "."))
        except ValueError:
            raise InvalidNumericInputException("Digite um numero valido.")
